File: src/lib/exportapi.test.ts
import { describe, expect, it } from 'vitest'
import {
  buildExportNdjson,
  EXPORT_FROM_NOT_FOUND,
  EXPORT_LIMIT_ERROR,
  exportContentDisposition,
  exportFilename,
  formatExportNotifyMessage,
  formatExportTimestamp,
  parseExportRecordsParams,
} from '@/lib/exportapi'
import { INVALID_RECORD_ID, type Record } from '@/lib/record'
import { serializeRecord } from '@/lib/recordjsonl'

const sample: Record = {
  id: '01900000-0000-7000-8000-000000000001',
  happened_at: '2026-07-30T00:00:00.000Z',
  numeric_value: '1.5',
  raw_content: null,
  tags: ['weight'],
  objective_context: 'scale',
  subjective_interpretation: null,
}

describe('parseExportRecordsParams', () => {
  it('requires limit in 1..1000', () => {
    expect(parseExportRecordsParams(new URLSearchParams())).toEqual({
      error: EXPORT_LIMIT_ERROR,
    })
    expect(
      parseExportRecordsParams(new URLSearchParams('limit=')),
    ).toEqual({ error: EXPORT_LIMIT_ERROR })
    expect(
      parseExportRecordsParams(new URLSearchParams('limit=0')),
    ).toEqual({ error: EXPORT_LIMIT_ERROR })
    expect(
      parseExportRecordsParams(new URLSearchParams('limit=1001')),
    ).toEqual({ error: EXPORT_LIMIT_ERROR })
    expect(
      parseExportRecordsParams(new URLSearchParams('limit=abc')),
    ).toEqual({ error: EXPORT_LIMIT_ERROR })
    expect(
      parseExportRecordsParams(new URLSearchParams('limit=1.5')),
    ).toEqual({ error: EXPORT_LIMIT_ERROR })
  })

  it('accepts limit without from', () => {
    expect(
      parseExportRecordsParams(new URLSearchParams('limit=100')),
    ).toEqual({ from: null, limit: 100 })
    expect(
      parseExportRecordsParams(new URLSearchParams('limit=1')),
    ).toEqual({ from: null, limit: 1 })
    expect(
      parseExportRecordsParams(new URLSearchParams('limit=1000')),
    ).toEqual({ from: null, limit: 1000 })
  })

  it('rejects invalid from UUID with Invalid record id', () => {
    expect(
      parseExportRecordsParams(
        new URLSearchParams('from=not-a-uuid&limit=10'),
      ),
    ).toEqual({ error: INVALID_RECORD_ID })
    expect(
      parseExportRecordsParams(
        new URLSearchParams(
          'from=01234567-89ab-cdef-0123-456789abcdef&limit=10',
        ),
      ),
    ).toEqual({ error: INVALID_RECORD_ID })
  })

  it('accepts valid from + limit', () => {
    expect(
      parseExportRecordsParams(
        new URLSearchParams(
          'from=01900000-0000-7000-8000-000000000001&limit=50',
        ),
      ),
    ).toEqual({
      from: '01900000-0000-7000-8000-000000000001',
      limit: 50,
    })
  })
})

describe('buildExportNdjson / filename / notify', () => {
  it('builds empty body for 0 rows', () => {
    expect(buildExportNdjson([])).toBe('')
  })

  it('builds NDJSON with trailing newline per row', () => {
    const second: Record = {
      ...sample,
      id: '01900000-0000-7000-8000-000000000002',
    }
    const body = buildExportNdjson([sample, second])
    expect(body).toBe(
      `${serializeRecord(sample)}\n${serializeRecord(second)}\n`,
    )
  })

  it('formats filename and disposition', () => {
    const now = new Date('2026-08-03T08:41:00.123Z')
    expect(formatExportTimestamp(now)).toBe('20260803T084100Z')
    expect(exportFilename(null, 100, now)).toBe(
      'records-from-start-limit-100-20260803T084100Z.jsonl',
    )
    expect(
      exportFilename('01900000-0000-7000-8000-000000000001', 50, now),
    ).toBe(
      'records-from-01900000-0000-7000-8000-000000000001-limit-50-20260803T084100Z.jsonl',
    )
    expect(exportContentDisposition(null, 10, now)).toBe(
      'attachment; filename="records-from-start-limit-10-20260803T084100Z.jsonl"',
    )
  })

  it('formats notify message', () => {
    expect(formatExportNotifyMessage(0, null, 100)).toBe(
      'Exported 0 records (from start, limit 100)',
    )
    expect(
      formatExportNotifyMessage(
        3,
        '01900000-0000-7000-8000-000000000001',
        50,
      ),
    ).toBe(
      'Exported 3 records (from 01900000-0000-7000-8000-000000000001, limit 50)',
    )
  })

  it('exports from-not-found constant', () => {
    expect(EXPORT_FROM_NOT_FOUND).toBe('export from id not found')
  })
})

File: faas/internal/exportapi/exportapi_test.go
package exportapi

import (
	"net/url"
	"testing"
	"time"

	"github.com/mdk/digitaltwin2026/faas/internal/record"
	"github.com/mdk/digitaltwin2026/faas/internal/recordjsonl"
)

func TestParseExportRecordsParams(t *testing.T) {
	_, err := ParseExportRecordsParams(url.Values{})
	if err == nil || err.Message != ErrExportLimitError {
		t.Fatalf("missing limit: %v", err)
	}
	_, err = ParseExportRecordsParams(url.Values{"limit": {"0"}})
	if err == nil || err.Message != ErrExportLimitError {
		t.Fatalf("limit 0: %v", err)
	}
	_, err = ParseExportRecordsParams(url.Values{"limit": {"1001"}})
	if err == nil || err.Message != ErrExportLimitError {
		t.Fatalf("limit 1001: %v", err)
	}
	_, err = ParseExportRecordsParams(url.Values{"limit": {"abc"}})
	if err == nil || err.Message != ErrExportLimitError {
		t.Fatalf("limit abc: %v", err)
	}

	p, err := ParseExportRecordsParams(url.Values{"limit": {"100"}})
	if err != nil || p.From != "" || p.Limit != 100 {
		t.Fatalf("got %+v err %v", p, err)
	}

	_, err = ParseExportRecordsParams(url.Values{"from": {"not-a-uuid"}, "limit": {"10"}})
	if err == nil || err.Message != record.ErrInvalidID {
		t.Fatalf("invalid from: %v", err)
	}

	id := "01900000-0000-7000-8000-000000000001"
	p, err = ParseExportRecordsParams(url.Values{"from": {id}, "limit": {"50"}})
	if err != nil || p.From != id || p.Limit != 50 {
		t.Fatalf("got %+v err %v", p, err)
	}
}

func TestBuildExportNdjsonAndFilename(t *testing.T) {
	empty, me := BuildExportNdjson(nil)
	if me != nil || empty != "" {
		t.Fatalf("empty: %q %v", empty, me)
	}

	num := "1.5"
	rec := record.Record{
		ID:               "01900000-0000-7000-8000-000000000001",
		HappenedAt:       "2026-07-30T00:00:00.000Z",
		NumericValue:     &num,
		RawContent:       nil,
		Tags:             []string{"weight"},
		ObjectiveContext: "scale",
		AiAnalysis:       nil,
	}
	line, err := recordjsonl.SerializeRecord(rec)
	if err != nil {
		t.Fatal(err)
	}
	body, me := BuildExportNdjson([]record.Record{rec})
	if me != nil {
		t.Fatal(me)
	}
	if body != line+"\n" {
		t.Fatalf("body %q want %q", body, line+"\n")
	}

	now := time.Date(2026, 8, 3, 8, 41, 0, 123000000, time.UTC)
	if got := ExportFilename("", 100, now); got != "records-from-start-limit-100-20260803T084100Z.jsonl" {
		t.Fatalf("filename start: %s", got)
	}
	if got := ExportFilename(rec.ID, 50, now); got != "records-from-01900000-0000-7000-8000-000000000001-limit-50-20260803T084100Z.jsonl" {
		t.Fatalf("filename from: %s", got)
	}
	if got := ExportContentDisposition("", 10, now); got != `attachment; filename="records-from-start-limit-10-20260803T084100Z.jsonl"` {
		t.Fatalf("disposition: %s", got)
	}
	if got := FormatExportNotifyMessage(0, "", 100); got != "Exported 0 records (from start, limit 100)" {
		t.Fatalf("notify: %s", got)
	}
	if got := FormatExportNotifyMessage(3, rec.ID, 50); got != "Exported 3 records (from 01900000-0000-7000-8000-000000000001, limit 50)" {
		t.Fatalf("notify from: %s", got)
	}
	if ErrExportFromNotFound != "export from id not found" {
		t.Fatal(ErrExportFromNotFound)
	}
}
